Lib/test/test_threading.py

```diff
@@ -1171,6 +1171,41 @@ def __del__(self):
         self.assertEqual(out.strip(), b"OK")
         self.assertIn(b"can't create new thread at interpreter shutdown", err)
 
+    @unittest.skipIf(support.Py_GIL_DISABLED, "gh-124149: daemon threads don't force exit")
+    def test_join_force_terminated_daemon_thread_in_finalization(self):
+        # gh-123940: Py_Finalize() forces all daemon threads to exit
+        # immediately (without unwinding the stack) upon acquiring the
+        # GIL. Finalizers that run after this must be able to join the daemon
+        # threads that were forced to exit.
```

Comment:
We wait for all non-daemon threads to finish as one of the first steps during finalization.  Thus there shouldn't be any non-daemon threads joining daemon threads at the point daemon threads start being killed.  If that actually is happening then something else is broken.  Otherwise, that means we're dealing with daemon threads joining daemon threads.  All of those should be killed during finalization, so I'm unclear on where we're running into a problem to be solved.

FWIW, we should probably emit a warning if a non-daemon thread ever tries to join a daemon thread.

Reply:
Ah, we're joining in the main thread in an object finalizer.  That makes sense.

Reply:
That said, yuck!

Reply:
> That said, yuck!

Agreed!